boot/kernel_os.go

```diff
@@ -20,80 +20,77 @@
 package boot
 
 import (
-	"errors"
 	"fmt"
 	"path/filepath"
-	"strings"
 
 	"github.com/snapcore/snapd/bootloader"
 	"github.com/snapcore/snapd/logger"
-	"github.com/snapcore/snapd/release"
 	"github.com/snapcore/snapd/snap"
 )
 
-// RemoveKernelAssets removes the unpacked kernel/initrd for the given
-// kernel snap.
-func RemoveKernelAssets(s snap.PlaceInfo) error {
+type coreBootParticipant struct {
+	s snap.PlaceInfo
+	t snap.Type
+}
+
+type coreKernel struct {
+	s snap.PlaceInfo
+}
+
+var _ BootParticipant = (*coreBootParticipant)(nil)
+var _ BootParticipant = (*coreKernel)(nil)
+
+func (k *coreKernel) RemoveKernelAssets() error {
+	// XXX: shouldn't we check the snap type?
 	bootloader, err := bootloader.Find()
 	if err != nil {
-		return fmt.Errorf("no not remove kernel assets: %s", err)
+		return fmt.Errorf("cannot remove kernel assets: %s", err)
 	}
 
 	// ask bootloader to remove the kernel assets if needed
-	return bootloader.RemoveKernelAssets(s)
+	return bootloader.RemoveKernelAssets(k.s)
 }
 
-// ExtractKernelAssets extracts kernel/initrd/dtb data from the given
-// kernel snap, if required, to a versioned bootloader directory so
-// that the bootloader can use it.
-func ExtractKernelAssets(s *snap.Info, snapf snap.Container) error {
-	if s.GetType() != snap.TypeKernel {
-		return fmt.Errorf("cannot extract kernel assets from snap type %q", s.GetType())
-	}
-
+func (k *coreKernel) ExtractKernelAssets(snapf snap.Container) error {
```

Comment:
the order of these definitions vs the struct definitions is a bit atypical, unless I'm confused